waiter/src/waiter/auth/oidc.clj

```diff
@@ -0,0 +1,252 @@
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns waiter.auth.oidc
+  (:require [clj-time.coerce :as tc]
+            [clj-time.core :as t]
+            [clojure.core.async :as async]
+            [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [waiter.auth.authentication :as auth]
+            [waiter.auth.jwt :as jwt]
+            [waiter.cookie-support :as cookie-support]
+            [waiter.status-codes :refer :all]
+            [waiter.util.http-utils :as hu]
+            [waiter.util.ring-utils :as ru]
+            [waiter.util.utils :as utils])
+  (:import (java.net URI)
+           (java.security SecureRandom)))
+
+(def ^:const accept-redirect-header-name "accept-redirect")
+
+(def ^:const accept-redirect-auth-header-name "accept-redirect-auth")
+
+(def ^:const challenge-cookie-duration-secs 60)
+
+(def ^:const code-verifier-length 128)
+
+(def ^:const content-security-policy-value "default-src 'none'; frame-ancestors 'none'")
+
+(def ^:const oidc-challenge-cookie "x-waiter-oidc-challenge")
+
+(def ^:const oidc-callback-uri "/oidc/v1/callback")
+
+(let [allowed-chars "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"
+      allowed-chars-count (count allowed-chars)
+      secure-rng (SecureRandom.)]
+  (defn create-code-verifier
+    "Creates a randomly generated string of length code-verifier-length from the list of allowed characters."
+    []
+    (apply str (repeatedly code-verifier-length #(nth allowed-chars (.nextInt secure-rng allowed-chars-count))))))
+
+(defn create-state-code
+  "Creates an encoded string of the input state map."
+  [state-map password]
+  (utils/map->base-64-string state-map password))
+
+(defn parse-state-code
+  "Parses the encoded string into the state map."
+  [state-str password]
+  (utils/base-64-string->map state-str password))
+
+(defn validate-oidc-callback-request
+  [password {:keys [headers] :as request}]
+  (let [{:strs [code state]} (-> request ru/query-params-request :query-params)
+        challenge-cookie (some-> headers
+                           (get "cookie")
+                           (cookie-support/cookie-value oidc-challenge-cookie))
+        bad-request-map {:log-level :info
+                         :query-param-state state
+                         :status http-400-bad-request}]
+    (when (str/blank? code)
+      (throw (ex-info "Query parameter code is missing" bad-request-map)))
+    (when (str/blank? state)
+      (throw (ex-info "Query parameter state is missing" bad-request-map)))
+    (when (str/blank? challenge-cookie)
+      (throw (ex-info "No challenge cookie set" bad-request-map)))
+    (let [state-map (try
+                      (-> state
+                        ;; handle url decoding of + to whitespace
+                        (str/replace #" " "+")
```

Comment:
Something is still off here.  `state` shouldn't have been decoded to contain a space.

Reply:
For reference, `java.util.Base64` decoding works without expanding the `+` to space:
```
(let [original "a=b+c&d=e=="
      enc-str (.encodeToString (Base64/getUrlEncoder) (.getBytes original))
      dec-str (String. (.decode (Base64/getDecoder) enc-str))]
  (println "enc-str:" enc-str)
  (println "dec-str:" dec-str))
enc-str: YT1iK2MmZD1lPT0=
dec-str: a=b+c&d=e==
```

Reply:
Avoids `ring-params`, instead added code to create the query parameter map by parsing the query string to avoid the white space issue.